pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -82,147 +86,222 @@
-
   // Cache expire time for ignored segment if there is no update from the segment.
   private static final int IGNORED_SEGMENT_CACHE_TIME_MINUTES = 10;
-
-  // Per partition info for all partitions active for the current table.
-  private final Map<Integer, IngestionInfo> _ingestionInfoMap = new ConcurrentHashMap<>();
-
-  // We mark partitions that go from CONSUMING to ONLINE in _partitionsMarkedForVerification: if they do not
-  // go back to CONSUMING in some period of time, we verify whether they are still hosted in this server by reading
-  // ideal state. This is done with the goal of minimizing reading ideal state for efficiency reasons.
-  // TODO: Consider removing this mechanism after releasing 1.2.0, and use {@link #stopTrackingPartitionIngestionDelay}
-  //       instead.
-  private final Map<Integer, Long> _partitionsMarkedForVerification = new ConcurrentHashMap<>();
-
-  private final Cache<String, Boolean> _segmentsToIgnore =
-      CacheBuilder.newBuilder().expireAfterAccess(IGNORED_SEGMENT_CACHE_TIME_MINUTES, TimeUnit.MINUTES).build();
-
-  // TODO: Make thread pool a server/cluster level config
-  // ScheduledExecutorService to check partitions that are inactive against ideal state.
-  private final ScheduledExecutorService _scheduledExecutor = Executors.newScheduledThreadPool(2);
+  // Timeout after 5 seconds while fetching the latest stream offset.
+  private static final long LATEST_STREAM_OFFSET_FETCH_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(5);
 
   private final ServerMetrics _serverMetrics;
   private final String _tableNameWithType;
   private final String _metricName;
-
   private final RealtimeTableDataManager _realTimeTableDataManager;
   private final Supplier<Boolean> _isServerReadyToServeQueries;
+  private final Cache<String, Boolean> _segmentsToIgnore =
+      CacheBuilder.newBuilder().expireAfterAccess(IGNORED_SEGMENT_CACHE_TIME_MINUTES, TimeUnit.MINUTES).build();
+  // Map to describe the partitions for which the metrics are being reported.
+  private final Map<Integer, Boolean> _partitionsTracked = new ConcurrentHashMap<>();
+  // Map to hold the ingestion info reported by the consumer.
+  private final Map<Integer, IngestionInfo> _ingestionInfoMap = new ConcurrentHashMap<>();
+  private final Map<Integer, Long> _partitionsMarkedForVerification = new ConcurrentHashMap<>();
+  private Clock _clock = Clock.systemUTC();
+  private final ScheduledExecutorService _metricsCleanupScheduler;
+  private final ScheduledExecutorService _ingestionDelayTrackingScheduler;
 
-  private Clock _clock;
+  // List of StreamMetadataProvider to fetch upstream latest stream offset (List because table can have multiple
+  // upstream topics)
+  protected List<StreamMetadataProvider> _streamMetadataProviderList;
+  protected volatile Map<Integer, StreamPartitionMsgOffset> _partitionIdToLatestOffset;
+  protected volatile Set<Integer> _partitionsHostedByThisServer = new HashSet<>();
```

Comment:
Should this set be synchronized since it is access by both tracking and cleanup?